File: src/sort/RadixSort.java
package sort;

import java.util.Arrays;
import java.util.Comparator;
import java.util.stream.IntStream;

public class RadixSort {
    /**
     * 基数排序
     *
     * @param arr 需要排序的数组
     */
    public static void sort(Integer[] arr) {
        if (arr.length < 2) return;
        radixSort(arr, 0, arr.length - 1, maxBits(arr));
    }

    /**
     * 将数组[l..r]区间排有序
     *
     * @param arr     数组
     * @param l       左边界
     * @param r       右边界
     * @param maxBits 数组最大值的位数
     */
    private static void radixSort(Integer[] arr, int l, int r, int maxBits) {
        Integer[] temp = new Integer[r - l + 1];
        for (int d = 1; d <= maxBits; d++) {
            int[] count = new int[10];
            // 遍历[l..r]区间上, 从右向左第d位的数, 将他对count的值 +1
            for (int i = l; i <= r; i++) {
                count[getDigit(arr[i], d)]++;
            }
            // 前缀和
            for (int i = 1; i < count.length; i++) {
                count[i] += count[i - 1];
            }
            // 从后向前遍历, 我们可以保证后面的一定出现在靠后的索引
            for (int i = r; i >= l; i--) {
                temp[--count[getDigit(arr[i], d)]] = arr[i];
            }
            System.arraycopy(temp, 0, arr, l, temp.length);
        }
    }

    /**
     * 计算给定数字, 从右向左第d位的值
     *
     * @param num 数字
     * @return num从右向左第d位的值
     */
    private static int getDigit(int num, int d) {
        return num / (int) Math.pow(10, d - 1) % 10;
    }

    /**
     * 返回给定数组中, 最大值的位数
     *
     * @param arr 数组
     * @return 数组中最大值的位数
     */
    private static int maxBits(Integer[] arr) {
        int max = Arrays.stream(arr).max(Comparator.naturalOrder()).get();
        int maxBits = 0;
        while (max != 0) {
            max /= 10;
            maxBits++;
        }
        return maxBits;
    }

    public static void main(String[] args) {
        Integer[] arr = IntStream.generate(() -> (int) (Math.random() * 1000))
                .limit(100)
                .boxed()
                .toArray(Integer[]::new);
        Integer[] std = arr.clone();
        Arrays.sort(std);
        sort(arr);
        System.out.println(Arrays.equals(std, arr));
    }
}

File: src/sort/radix_sort.py
from typing import List
import sort_util


def find_max_bits(arr: List[int]) -> int:
    """返回给定数组最大数的位数"""
    max_val = max(arr)
    bits = 0
    while max_val > 0:
        max_val //= 10
        bits += 1
    return bits


def get_digit(num: int, i: int) -> int:
    """获取num的第i位的数, i从0索引开始"""
    return num // 10 ** i % 10


def radix_sort_process(arr: List[int], l: int, r: int, max_bits: int):
    """基数排序将[l,r]范围排有序, 其中最数的位数为max_bits
    :param arr: 需要排序的数组
    :param l: 左边界
    :param r: 右边界
    :param max_bits: 数组中最大数的位数
    """
    temp = [0] * (r - l + 1)
    # 遍历最大位数次
    for i in range(max_bits):
        # 创建0..10基数数组
        radix_arr = [0] * 10
        for num in arr:
            radix_arr[get_digit(num, i)] += 1
        # 构建前缀和数组
        for j in range(1, len(radix_arr)):
            radix_arr[j] += radix_arr[j - 1]
        # 反向迭代数组, 获取num对应前缀和数组中的我位置
        for j in range(r, l - 1, -1):
            radix_arr[get_digit(arr[j], i)] -= 1
            temp[radix_arr[get_digit(arr[j], i)]] = arr[j]
        arr[l:r + 1] = temp


def radix_sort(arr: List[int]):
    """基数排序
    """
    radix_sort_process(arr, 0, len(arr) - 1, find_max_bits(arr))


if __name__ == '__main__':
    sort_util.check(radix_sort)
